qiskit/chemistry/ground_state_calculation/mes_ground_state_calculation.py

```diff
@@ -0,0 +1,84 @@
+#
+# (C) Copyright IBM 2020.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Ground state computation using a minimum eigensolver."""
+
+from typing import Union
+
+from qiskit.aqua.algorithms import MinimumEigensolver
+from qiskit.chemistry.drivers import BaseDriver
+from qiskit.chemistry.ground_state_calculation import GroundStateCalculation
+from qiskit.chemistry.qubit_transformations import QubitOperatorTransformation
+from qiskit.chemistry.results import EigenstateResult
+
+from .mes_factories import MESFactory
+
+
+class MinimumEigensolverGroundStateCalculation(GroundStateCalculation):
+    """Ground state computation using a minimum eigensolver."""
+
+    def __init__(self, transformation: QubitOperatorTransformation,
+                 solver: Union[MinimumEigensolver, MESFactory]) -> None:
+        """
+
+        Args:
+            transformation: Qubit Operator Transformation
+            solver: Minimum Eigensolver or MESFactory object, e.g. the VQEUCCSDFactory.
+        """
+        super().__init__(transformation)
+        self._solver = solver
+
+    @property
+    def solver(self) -> Union[MinimumEigensolver, MESFactory]:
+        """Returns the minimum eigensolver or factory."""
+        return self._solver
+
+    @solver.setter
+    def solver(self, solver: Union[MinimumEigensolver, MESFactory]) -> None:
+        """Sets the minimum eigensolver or factory."""
+        self._solver = solver
+
+    def returns_groundstate(self) -> bool:
+        """TODO
+        whether the eigensolver returns the ground state or only ground state energy."""
+
+        return False
+
+    def compute_groundstate(self, driver: BaseDriver) -> EigenstateResult:
+        """Compute Ground State properties.
+
+        Args:
+            driver: a chemistry driver object which defines the chemical problem that is to be
+                    solved by this calculation.
+
+        Returns:
+            An eigenstate result. Depending on the transformation this can be an electronic
+            structure or bosonic result.
+        """
+        operator, aux_operators = self.transformation.transform(driver)
+
+        if isinstance(self._solver, MESFactory):
+            # this must be called after transformation.transform
+            solver = self._solver.get_solver(self.transformation)
+        else:
+            solver = self._solver
+
+        aux_operators = aux_operators if solver.supports_aux_operators() else None
+
+        raw_mes_result = solver.compute_minimum_eigenvalue(operator, aux_operators)
+
+        eigenstate_result = EigenstateResult()
```

Comment:
I guess I am trying the understand the utility of EigenstateResult. The minumum eigen solver result already has these fields so I am not sure why they need to be copied for result to be interpreted by the transform. The Hamiltonian class directly took the minimum eigen solver result, and interpreted it for a chemistry outcome. This meant anyone could use the Hamiltonian get the operator/aux_ops and run VQE or whatever themselves. They would then call the interpret on the VQE result to translate it back. Now this seems to have another translation step there before the Transform can interpret it. It seems to be more prohibitive to using the blocks we have in way the user can use/combine more freely

Reply:
The reason for the introduction of the `EigenstateResult` is that it is supposed to work with both, ground and excited state calculations. The respective `EigensolverResult` and `MinimumEigensolverResult` classes cannot be used for this since they do not have a common super-type.
I think this is not a problem of this interface but rather a problem on the `Eigensolver`/`MinimumEigensolver` interface. But this is a problem for another time and shouldn't be part of this PR.

Reply:
The difficulty for me is that interpret on the Hamiltonian would work with the ground state or excited states results from Aqua. Yes it handled both. This allows a user to use the Hamiltonian - or the Transformation - get the output operators, potentially manipulate them if they wish, and call VQE. The to get the chemistry result give the Aqua (VQE) result to the Hamiltonian so it can process it and given back a complete result. Maybe the interpret should take either like it did before. Having it interpret something else that is built out of the MinEigResult - if that building out is done in a standard way, if its needed, it could be done in the interpret too. Maybe I am thinking too much of the Aqua result objects here and I do not know what the intersection is with the new excited state logic yet not having seen it. Either way I think we need a simple way to pass in the Aqua (Min)EigSolver result and get a chemistry result like we had in the past.

Reply:
There are a few reasons why we are strongly opinionated about keeping the `EigenstateResult` even though it appears to just provide another layer between the final result and the `(Minimum)EigensolverResult`:
1. We want to be able to reuse this class for both, ground and excited states. This is not possible in the case of the other two classes and this would lead to at least four different final types: `ElectronicGSResult`, `ElectronicESResult`, `VibronicGSResult`, and `VibronicESResult` 
2. We need to enforce the `raw_result` field prior to the final types. Since neither of `(Min)EigensolverResult` provide this field, we want to do so in a super type.
3. And as you pointed out yourself, you are thinking too much about the Aqua result objects. By cleanly separating the chemistry results from those (even if we mimic them) we have better separation between the modules.
4. In case that the `(Min)EigensolverResult` classes happen to match our requirements at some point in the future we can still go back to using them.

Reply:
I don't have a problem with chemistry having its own results. What would be nice is that I can use the blocks in chemistry at different granularities simply enough. FermionicTransformation emits qubit op etc. like the Hamiltonian before. Users can take the qubit op and call the Aqua algorithms themselves. There is now however an impedance mismatch - before one could take the Aqua result and call interpret on the Hamiltonian to get a Chemistry result. Now there is some creation of this EigenstateResult before things can get interpreted. To me it would make sense that interpret can take a MigEigResult - maybe as one of a set of types it can take along with an EigResult, and if it needs 'translating' into an EigenstateResult do it there.

Reply:
I just extended the `interpret()` method to also handle `(Minimum)EigensolverResult`s as discussed offline.
This also allowed moving the construction of the chemistry-stack specific `EigenstateResult` object into that method which makes the implementation inside of the calculation classes much cleaner.